### PR TITLE
Json fixes

### PR DESCRIPTION
**Pull request by nicolasnoble**
Fixing two new cases reported by @StilesCrisis.

-) EOF should properly emit remaining tokens.
-) GRPC_JSON_STATE_VALUE_BEGIN is not a valid state to end an object with, obviously.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +2.6%     +68 src/core/lib/json/json_reader.cc     +68  +2.6%
      +2.7%     +68 grpc_json_reader_run                 +68  +2.7%

 -------------- SHRINKING                        --------------
  -0.0%      -4 [None]                               -60  -0.0%

  +0.0%     +64 TOTAL                                 +8  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/lib/json/json_reader.cc [hunk 1]
```diff
@@ -147,6 +142,12 @@ grpc_json_reader_status grpc_json_reader_run(grpc_json_reader* reader) {
           default:
             return GRPC_JSON_PARSE_ERROR;
         }
+        if (c != GRPC_JSON_READ_CHAR_EOF) break;
```
Comment (StilesCrisis) [inline src/core/lib/json/json_reader.cc]: change to

if (c == GRPC_JSON_READ_CHAR_EOF) {

File: test/core/json/json_test.cc [hunk 2]
```diff
@@ -107,6 +107,8 @@ static testing_pair testing_pairs[] = {
     {"", nullptr},
     /* Testing extra characters after end of parsing. */
     {"{},", nullptr},
+    /* Testing bad containers. */
+    {"{\"x\":}", nullptr},
```
Comment (StilesCrisis) [inline test/core/json/json_test.cc]: move to line 121 to be with the existing bad-container tests

Comment (StilesCrisis) [reply test/core/json/json_test.cc]: Also, please add a test for {"x":0,"y":}

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +2.6%     +68 src/core/lib/json/json_reader.cc     +68  +2.6%
      +2.7%     +68 grpc_json_reader_run                 +68  +2.7%

 -------------- SHRINKING                        --------------
  -0.0%      -4 [None]                               -60  -0.0%

  +0.0%     +64 TOTAL                                 +8  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +2.6%     +68 src/core/lib/json/json_reader.cc     +68  +2.6%
      +2.7%     +68 grpc_json_reader_run                 +68  +2.7%

 -------------- SHRINKING                        --------------
  -0.0%      -4 [None]                               -60  -0.0%

  +0.0%     +64 TOTAL                                 +8  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +2.6%     +68 src/core/lib/json/json_reader.cc     +68  +2.6%
      +2.7%     +68 grpc_json_reader_run                 +68  +2.7%

 -------------- SHRINKING                        --------------
  -0.0%      -4 [None]                               -60  -0.0%

  +0.0%     +64 TOTAL                                 +8  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                        FILE SIZE
 ++++++++++++++ GROWING                          ++++++++++++++
  +2.6%     +68 src/core/lib/json/json_reader.cc     +68  +2.6%
      +2.7%     +68 grpc_json_reader_run                 +68  +2.7%
  +0.9%     +16 src/core/lib/json/json_string.cc     +16  +0.9%
       +28%     +24 json_reader_string_clear             +24   +28%

 -+-+-+-+-+-+-+ MIXED                            +-+-+-+-+-+-+-
  +0.0%     +12 [None]                               -68  -0.0%

  +0.0%     +96 TOTAL                                +16  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by StilesCrisis** [comment]
Bumping issue

**Pull request closed without merging**